JIT: initial profile repair

AndyAyersMS opened the pull request:
Always run a repair pass after incorporating profile data. If that fails to converge, blend in synthetic likelihoods in increasing strengths to enable convergence (giving up some accuracy).

Enable block weight consistency checks to run right after profile incorporation. Disable just after that.

We can generally make profile data self-consistent, save for two exceptions:
1. if the flow graph contains an infinite (or effectively infinite) loop
2. if the IL is invalid

Because of (2) we can't immediately assert if the profile is not consistent; we must defer until after we've successfully imported the method (since the importer contains many of the validity checks).

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=635050&view=ms.vss-build-web.run-extensions-tab)

AndyAyersMS committed:
JIT: intial profild repair

Always run a repair pass after incorporating profile data. If that fails
to converge, blend in synthetic likelihoods in increasing strengths to
enable convergence (giving up some accuracy).

Enable block weight consistency checks to run right after profile incorporation.
Disable just after that.

AndyAyersMS commented:
@amanasifkhalid FYI

This may be a little too drastic....

AndyAyersMS commented:
```
'System.Buffers.AhoCorasick:IndexOfAnyCore[System.Buffers.StringSearchValuesHelper+CaseSensitive,System.Buffers.AhoCorasick+NoFastScan](System.ReadOnlySpan`1[ushort]):int:this' during 'Profile incorporation' (IL size 388; hash 0x646e0c1f; Tier1)
```
This is the same method that inspired https://github.com/dotnet/runtime/pull/100449.

AndyAyersMS commented:
This shows we can have initially consistent profiles (with one exception) but perhaps more TP or CQ/CS impact than is necessary.

Because this method perturbs the profile there are several aspects to the TP impact: the cost of the computation, the effects of that profile on the rest of the JIT, and the possibility of additional missed contexts.

I will work up a version that does all the computation but then doesn't actually modify the profile, so I can get a better read on the first part. If the computation is relatively cheap it may make sense to up the solver iteration limit and/or slow down the blend factor growth "schedule" so the repairs are less impactful overall (in other words: perhaps by doing more computation early we can end up with less TP/CS/CQ churn).

Note we only expect to have to iterate if there are one or more "rare" factors:
* improper (irreducible) loop headers
* capped cyclic probabilities
* infinite (or effectively infinite) loops (note these can never converge)
* extremely large counts

So for the vast majority of methods we should do one pass through the code and be done. That's something I also should measure....

amanasifkhalid commented:
> This shows we can have initially consistent profiles (with one exception) but perhaps more TP or CQ/CS impact than is necessary.

Out of curiosity, are you planning on using the SPMI results to determine if the tradeoff between profile accuracy and profile consistency is worth it here? You mentioned you plan to try to reduce churn, so I might be premature in looking at the diffs, but I see some of the largest size changes are due to loop cloning, so those methods that regressed might be perf wins (and the size improvements might not be ideal?). As for the large number of small regressions that seem to be from block ordering, it doesn't seem easy to tell if splitting up some fall-through is an improvement because it allowed us to move a hot block up, or a regression because the newly-added jump isn't cold enough to pay for itself (though maybe spending a nontrivial amount of time on these diffs isn't worth it if we're going to change the block layout algorithm soon).

AndyAyersMS committed:
allow infinite rel residual at any stage; better logging of loops containing improper loops

AndyAyersMS commented:
> Out of curiosity, are you planning on using the SPMI results to determine if the tradeoff between profile accuracy and profile consistency is worth it here? 

Yes, to some extent... I think it might make sense to insist on initial consistency for now and later when we see how quickly consistency degrades (or is maintained) we might redo things. Maybe.

AndyAyersMS committed:
Fix some issues validating entry/exit balance for methods with throws.

Synthesis and reconstruction assume exceptions are rare, so if the actual
profile shows significant flow into a throw that can be caught, we won't
model this properly. The net result is that some profile weight will
vanish, throwing off the entry/exit balance.

Tolerate this for now, by watching for cachable throws, and disabling the
entry/exit balance checks if any are seen. If it ever turns out that serious
code has high exception frequency we can reconsider and try and model flow
through catches.

AndyAyersMS committed:
fix gcc build issue; elaborate on the iterative blend strategy

AndyAyersMS committed:
Merge branch 'main' into InitialProfileRepair

AndyAyersMS commented:
/azp run runtime-coreclr libraries-pgo

AndyAyersMS commented:
TP diffs are still kind of high, suspect this is from knock-on effects where the new profile leads to inlining or cloning changes.

![image](https://github.com/dotnet/runtime/assets/10121823/6918bb0b-9913-4142-8e01-05d85ef30399)

Number of missed contexts is also up considerably, so the "tp win" on asp.net is likely misleading too.

Going to do some local runs where I just run the solver but don't update the counts, to see what that costs.

AndyAyersMS commented:
Can't repro the arm64 failure so far.

Getting an isolated TP solver measurement seems tricky... still working on it.

AndyAyersMS committed:
blend returns too; increase iteration limit

AndyAyersMS committed:
Merge remote-tracking branch 'upstream/main' into InitialProfileRepair

AndyAyersMS committed:
format

AndyAyersMS commented:
Upping iteration limit, hoping things converge on one of the earlier passes, so we do less blending overall, and so have something closer to the initial profile, so less TP impact and less churn. Seems like the solver is (relatively) cheap, though not sure I trust my local TP data. And we may do less work overall.

Warp heuristic return likelihoods upwards each retry.

AndyAyersMS commented:
Hmm, looks like a step in the wrong direction, which is odd... need to dig in.

AndyAyersMS committed:
Solver should check entry/exit balance too.

Unless there is a catchable throw.::

AndyAyersMS commented:
> Hmm, looks like a step in the wrong direction, which is odd... need to dig in.

Looks like many of the new failures were entry/exit balance checks: that is we try and ensure that the total BBJ_RETURN and select BBJ_THROW weight matches the entry weight. Even if each node locally converges to 0.01 tolerance, a chain of these small mismatches can add up to a greater mismatch.

So we now check for the entry/exit residual during solving and keep iterating if it is too high.

AndyAyersMS commented:
> Upping iteration limit, hoping things converge on one of the earlier passes, so we do less blending overall, and so have something closer to the initial profile, so less TP impact and less churn. Seems like the solver is (relatively) cheap, though not sure I trust my local TP data.
> 
> Warp heuristic return likelihoods upwards each retry.

Idea that more initial iterations will improve TP and lead to less churn is not holding up so far... 

![image](https://github.com/dotnet/runtime/assets/10121823/1506ae99-7ce4-407d-aba1-3048b370a56f)

AndyAyersMS committed:
handle slight overflow in entry/exit residual

AndyAyersMS committed:
fix build

AndyAyersMS commented:
/azp run runtime-coreclr libraries-pgo

AndyAyersMS commented:
Looks better, correctness wise... just one issue left perhaps for cases where the intitial profile is synthetic. TP/CQ still not where I'd like them to be, but I'm not sure they can be improved much.

AndyAyersMS commented:
/azp run runtime-coreclr jitstress, runtime-coreclr pgostress, runtime-coreclr pgo

AndyAyersMS committed:
back off on entry/exit checking a bit more; make sure to reset m_overflow

AndyAyersMS commented:
/azp run runtime-coreclr libraries-pgo, runtime-coreclr jitstress, runtime-coreclr pgostress, runtime-coreclr pgo

AndyAyersMS committed:
allow some inconsistency

AndyAyersMS committed:
tone down volume of stuff in jit dump a bit

AndyAyersMS commented:
/azp run runtime-coreclr libraries-pgo, runtime-coreclr jitstress, runtime-coreclr pgostress, runtime-coreclr pgo

AndyAyersMS committed:
fix deferred checking mechanism

AndyAyersMS commented:
@amanasifkhalid PTAL
cc @dotnet/jit-contrib 

Will have some TP impact and cause some CQ churn. Hoping to get most of the TP back if we can reign in cloning a bit.

AndyAyersMS committed:
fix build and a few comment typos

amanasifkhalid reviewed (approved):
LGTM, assuming you get the results you want in CI.

Thanks for all the iterative work on this!

AndyAyersMS commented on src/coreclr/jit/fgprofilesynthesis.cpp:
```diff
@@ -216,15 +277,17 @@ void ProfileSynthesis::AssignLikelihoodCond(BasicBlock* block)
 
     if (isTrueThrow != isFalseThrow)
     {
+        const weight_t throwLikelihood = 0;
```
Makes sense.

AndyAyersMS committed:
review feedback

AndyAyersMS commented:
/azp run runtime-coreclr libraries-pgo, runtime-coreclr jitstress, runtime-coreclr pgostress, runtime-coreclr pgo

AndyAyersMS commented:
The extra legs may not be 100% clean, though there was just one (related) failure in my last go round, so it's possible.

AndyAyersMS commented:
libraries-pgo has a few crashes. I don't think they are directly related to this change, though perhaps it is making some existing issue more prominent. I will investigate them separately. There has been a low background rate of these things for a while now.

DrewScoggins commented:
Regressions
Linux Arm64 Ampere: https://github.com/dotnet/runtime/issues/100923
@AndyAyersMS
